File: code/apps/pyobjc/Managed Software Center/Managed Software Center/passwdutil.py
# -*- coding: utf-8 -*-
#
#  passwdutil.py
#  Managed Software Center
#
#  Created by Greg Neagle on 4/18/17.
#  Copyright (c) 2018-2019 The Munki Project. All rights reserved.
#
'''Code to interact with the OpenDirectory framework'''


import OpenDirectory as OD
from Foundation import NSLog

def findODuserRecords(username, nodename='/Search'):
    '''Uses OpenDirectory methods to find user records for username'''
    mySession = OD.ODSession.defaultSession()
    if not mySession:
        NSLog('DS session error: no default session')
        return None
    searchNode, err = OD.ODNode.nodeWithSession_name_error_(
                                                mySession, nodename, None)
    if not searchNode:
        NSLog('DS node error: %s' % err)
        return None
    myQuery, err = OD.ODQuery.queryWithNode_forRecordTypes_attribute_matchType_queryValues_returnAttributes_maximumResults_error_(
            searchNode,
            OD.kODRecordTypeUsers,
            OD.kODAttributeTypeRecordName,
            OD.kODMatchEqualTo, 
            username, 
            OD.kODAttributeTypeAllAttributes,
            0, 
            None)
    results, err = myQuery.resultsAllowingPartial_error_(False, None)
    if not results:
        return None
    return results


def findODuserRecord(username, nodename='/Search'):
    '''Returns first record found for username.'''
    records = findODuserRecords(username, nodename)
    if records:
        return records[0]
    else:
        return None


def verifyPassword(username, password):
    '''Uses OpenDirectory methods to verify password for username'''
    result = False
    userRecord = findODuserRecord(username)
    if userRecord:
        result, unused_err = userRecord.verifyPassword_error_(
                                                            password, None)
    return result

File: code/apps/Managed Software Center/Managed Software Center/passwdutil.swift
//
//  passwdutil.swift
//  Managed Software Center
//
//  Created by Greg Neagle on 6/29/18.
//  Copyright © 2018-2026 The Munki Project. All rights reserved.
//

import Foundation
import OpenDirectory

func findODuserRecords(username: String, nodename: String = "/Search") throws -> [ODRecord] {
    // Uses OpenDirectory methods to find user records for username
    let searchNode = try ODNode(session: ODSession.default(), name: nodename)
    let query = try ODQuery(node: searchNode,
                            forRecordTypes: kODRecordTypeUsers,
                            attribute: kODAttributeTypeRecordName,
                            matchType: ODMatchType(kODMatchEqualTo),
                            queryValues: username,
                            returnAttributes: kODAttributeTypeAllAttributes,
                            maximumResults: 0)
    return (try query.resultsAllowingPartial(false) as! [ODRecord])
}

func findODuserRecord(username: String, nodename: String = "/Search") -> ODRecord? {
    // Returns first record found for username, or nil if not found
    do {
        let records = try findODuserRecords(username: username)
        if records.isEmpty {
            return nil
        }
        return records[0]
    } catch {
        return nil
    }
}

func verifyODPassword(username: String, password: String) -> Bool {
    // Uses OpenDirectory methods to verify password for username
    // returns false if invalid username or wrong password for valid username,
    // true otherwise
    if let userRecord = findODuserRecord(username: username) {
        do {
            try userRecord.verifyPassword(password)
            return true
        } catch {
            return false
        }
    }
    return false
}
